src/Compilers/CSharp/Portable/Symbols/ArrayTypeSymbol.cs

```diff
@@ -16,77 +16,93 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
     /// <summary>
     /// An ArrayTypeSymbol represents an array type, such as int[] or object[,].
     /// </summary>
-    internal sealed partial class ArrayTypeSymbol : TypeSymbol, IArrayTypeSymbol
+    internal abstract partial class ArrayTypeSymbol : TypeSymbol, IArrayTypeSymbol
     {
         private readonly TypeSymbol _elementType;
-        private readonly int _rank;
         private readonly NamedTypeSymbol _baseType;
-        private readonly ImmutableArray<NamedTypeSymbol> _interfaces;
         private readonly ImmutableArray<CustomModifier> _customModifiers;
 
-        /// <summary>
-        /// Create a new ArrayTypeSymbol.
-        /// </summary>
-        /// <param name="elementType">The element type of this array type.</param>
-        /// <param name="customModifiers">Custom modifiers for the element type of this array type.</param>
-        /// <param name="rank">The rank of this array type.</param>
-        /// <param name="declaringAssembly">The assembly "declaring"/using the array type.</param>
-        internal ArrayTypeSymbol(
+        private ArrayTypeSymbol(
+            TypeSymbol elementType,
+            NamedTypeSymbol array,
+            ImmutableArray<CustomModifier> customModifiers)
+        {
+            Debug.Assert((object)elementType != null);
+            Debug.Assert((object)array != null);
+
+            _elementType = elementType;
+            _baseType = array;
+            _customModifiers = customModifiers.NullToEmpty();
+        }
+
+        internal static ArrayTypeSymbol CreateCSharpArray(
```

Comment:
Looking at the implementation I understand what this is doing now but the name didn't make it immediately obvious.  I can't think of a better name off hand though.